File: util/math/src/main/java/dev/sweety/math/pool/ArrayPool.java
package dev.sweety.math.pool;

import java.util.ArrayDeque;
import java.util.concurrent.ConcurrentLinkedDeque;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Consumer;
import java.util.function.IntFunction;
import java.util.function.ToIntFunction;

/**
 * Pooling interface for reusable arrays of any type.
 *
 * <p>Two implementations are available via the factory methods:
 * <ul>
 *   <li>{@link #threadLocal} — per-thread ArrayDeque, zero contention.
 *   <li>{@link #shared} — ConcurrentLinkedDeque, safe for cross-thread use.
 *       Fixes the peek/pollFirst TOCTOU present in the previous implementation:
 *       {@code pollFirst()} is called directly; if the taken array is too small it is
 *       dropped (not returned to the pool) and a fresh one is allocated.
 * </ul>
 *
 * <p>Both variants accept any array type via {@code IntFunction<T>} (factory) and
 * {@code ToIntFunction<T>} (length extractor). Convenience factories for {@code byte[]},
 * {@code int[]}, and {@code float[]} are provided.
 */
public interface ArrayPool<T> {

    /**
     * Returns an array whose length is {@code >= minSize}.
     * The returned array may be larger than requested.
     */
    @Acquire
    T acquire(int minSize);

    /** Returns {@code arr} to the pool if it is within the acceptable size range. */
    @Release
    void release(T arr);

    // ========================== TYPED CONVENIENCE FACTORIES ==========================

    static ArrayPool<byte[]> threadLocalBytes(int defaultSize, int maxPerThread) {
        return threadLocal(byte[]::new, a -> a.length, defaultSize, maxPerThread);
    }

    static ArrayPool<int[]> threadLocalInts(int defaultSize, int maxPerThread) {
        return threadLocal(int[]::new, a -> a.length, defaultSize, maxPerThread);
    }

    static ArrayPool<long[]> threadLocalLongs(int defaultSize, int maxPerThread) {
        return threadLocal(long[]::new, a -> a.length, defaultSize, maxPerThread);
    }

    static ArrayPool<float[]> threadLocalFloats(int defaultSize, int maxPerThread) {
        return threadLocal(float[]::new, a -> a.length, defaultSize, maxPerThread);
    }

    static ArrayPool<double[]> threadLocalDoubles(int defaultSize, int maxPerThread) {
        return threadLocal(double[]::new, a -> a.length, defaultSize, maxPerThread);
    }

    static ArrayPool<byte[]> sharedBytes(int defaultSize, int maxPoolSize) {
        return shared(byte[]::new, a -> a.length, defaultSize, maxPoolSize);
    }

    static ArrayPool<int[]> sharedInts(int defaultSize, int maxPoolSize) {
        return shared(int[]::new, a -> a.length, defaultSize, maxPoolSize);
    }

    static ArrayPool<long[]> sharedLongs(int defaultSize, int maxPoolSize) {
        return shared(long[]::new, a -> a.length, defaultSize, maxPoolSize);
    }

    static ArrayPool<float[]> sharedFloats(int defaultSize, int maxPoolSize) {
        return shared(float[]::new, a -> a.length, defaultSize, maxPoolSize);
    }

    static ArrayPool<double[]> sharedDoubles(int defaultSize, int maxPoolSize) {
        return shared(double[]::new, a -> a.length, defaultSize, maxPoolSize);
    }

    // ========================== GENERIC FACTORIES ==========================

    static <T> ArrayPool<T> threadLocal(IntFunction<T> factory, ToIntFunction<T> length,
                                        int defaultSize, Consumer<T> onDiscard, int maxPerThread) {
        return new ThreadLocalImpl<>(factory, length, defaultSize, onDiscard, maxPerThread);
    }

    static <T> ArrayPool<T> threadLocal(IntFunction<T> factory, ToIntFunction<T> length,
                                        int defaultSize, int maxPerThread) {
        return threadLocal(factory, length, defaultSize, _ -> {
        }, maxPerThread);
    }

    static <T> ArrayPool<T> shared(IntFunction<T> factory, ToIntFunction<T> length,
                                   int defaultSize, int maxPoolSize) {
        return new SharedImpl<>(factory, length, defaultSize, maxPoolSize);
    }

    // ========================== IMPLEMENTATIONS ==========================

    final class ThreadLocalImpl<T> implements ArrayPool<T> {
        private final ThreadLocal<ArrayDeque<T>> pool = ThreadLocal.withInitial(ArrayDeque::new);
        private final IntFunction<T> factory;
        private final ToIntFunction<T> length;
        private final int defaultSize;
        private final Consumer<T> onDiscard;
        private final int maxPerThread;

        ThreadLocalImpl(IntFunction<T> factory, ToIntFunction<T> length,
                        int defaultSize, Consumer<T> onDiscard, int maxPerThread) {
            this.factory = factory;
            this.length = length;
            this.defaultSize = defaultSize;
            this.onDiscard = onDiscard;
            this.maxPerThread = maxPerThread;
        }

        @Override
        public T acquire(int minSize) {
            ArrayDeque<T> deque = pool.get();
            T arr = deque.poll();
            if (arr != null && length.applyAsInt(arr) >= minSize) return arr;
            return factory.apply(Math.max(defaultSize, minSize));
        }

        @Override
        public void release(T arr) {
            if (arr == null) return;
            int len = length.applyAsInt(arr);
            if (len < defaultSize / 2 || len > defaultSize * 2) {
                onDiscard.accept(arr);
                return;
            }
            ArrayDeque<T> deque = pool.get();
            if (deque.size() < maxPerThread) deque.push(arr);
            else onDiscard.accept(arr);
        }
    }

    final class SharedImpl<T> implements ArrayPool<T> {
        private final ConcurrentLinkedDeque<T> pool = new ConcurrentLinkedDeque<>();
        private final AtomicInteger count = new AtomicInteger();
        private final IntFunction<T> factory;
        private final ToIntFunction<T> length;
        private final int defaultSize;
        private final int maxPoolSize;

        SharedImpl(IntFunction<T> factory, ToIntFunction<T> length,
                   int defaultSize, int maxPoolSize) {
            this.factory = factory;
            this.length = length;
            this.defaultSize = defaultSize;
            this.maxPoolSize = maxPoolSize;
        }

        @Override
        public T acquire(int minSize) {
            // pollFirst directly — no peek/poll TOCTOU
            T arr = pool.pollFirst();
            if (arr != null) {
                count.decrementAndGet();
                if (length.applyAsInt(arr) >= minSize) return arr;
                // too small — drop it, allocate fresh
            }
            return factory.apply(Math.max(defaultSize, minSize));
        }

        @Override
        public void release(T arr) {
            if (arr == null) return;
            int len = length.applyAsInt(arr);
            if (count.get() < maxPoolSize && len >= defaultSize / 2 && len <= defaultSize * 2) {
                pool.offerFirst(arr);
                count.incrementAndGet();
            }
        }
    }
}

File: util/math/src/main/kotlin/dev/sweety/math/pool/ArrayPool.kt
package dev.sweety.math.pool

import java.util.concurrent.ConcurrentLinkedDeque
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Consumer
import java.util.function.Function
import java.util.function.IntFunction
import java.util.function.ToIntFunction
import kotlin.math.max

/**
 * Pooling interface for reusable arrays of any type.
 * 
 * 
 * Two implementations are available via the factory methods:
 * 
 *  * [.threadLocal] — per-thread ArrayDeque, zero contention.
 *  * [.shared] — ConcurrentLinkedDeque, safe for cross-thread use.
 * Fixes the peek/pollFirst TOCTOU present in the previous implementation:
 * `pollFirst()` is called directly; if the taken array is too small it is
 * dropped (not returned to the pool) and a fresh one is allocated.
 * 
 * 
 * 
 * Both variants accept any array type via `IntFunction<T>` (factory) and
 * `ToIntFunction<T>` (length extractor). Convenience factories for `byte[]`,
 * `int[]`, and `float[]` are provided.
 */
interface ArrayPool<T> {
    /**
     * Returns an array whose length is `>= minSize`.
     * The returned array may be larger than requested.
     */
    @Acquire
    fun acquire(minSize: Int): T

    /**
     * Returns `arr` to the pool if it is within the acceptable size range.
     */
    @Release
    fun release(arr: T)

    /**
     * Borrows an array of at least `minSize`, applies `fn`, releases it, yields result.
     */
    @Borrows
    fun <V> use(minSize: Int, fn: Function<T, V>): V {
        val arr = acquire(minSize)
        try {
            return fn.apply(arr)
        } finally {
            release(arr)
        }
    }

    @Borrows
    fun consume(minSize: Int, fn: Consumer<T>) {
        val arr = acquire(minSize)
        try {
            fn.accept(arr)
        } finally {
            release(arr)
        }
    }

    // ========================== IMPLEMENTATIONS ==========================
    class ThreadLocalPool<T> internal constructor(
        private val factory: IntFunction<T>, private val length: ToIntFunction<T>,
        private val defaultSize: Int, private val onDiscard: Consumer<T>?, private val maxPerThread: Int
    ) : ArrayPool<T> {
        private val pool: ThreadLocal<ArrayDeque<T>> = ThreadLocal.withInitial { ArrayDeque() }

        override fun acquire(minSize: Int): T {
            val deque = pool.get()
            // Fast path: head is large enough
            val head = deque.firstOrNull()
            if (head != null && length.applyAsInt(head) >= minSize) return deque.removeFirst()
            // Slow path: scan for first array that fits, leave the rest in place
            val it = deque.iterator()
            while (it.hasNext()) {
                val arr = it.next()
                if (length.applyAsInt(arr) >= minSize) {
                    it.remove()
                    return arr
                }
            }
            return factory.apply(max(defaultSize, minSize))
        }

        override fun release(arr: T) {
            if (arr == null) return
            val len = length.applyAsInt(arr)
            if (len < defaultSize / 2 || len > defaultSize * 2) {
                onDiscard?.accept(arr)
                return
            }
            val deque = pool.get()
            if (deque.size < maxPerThread) deque.addFirst(arr)
            else onDiscard?.accept(arr)
        }
    }

    class SharedPool<T> internal constructor(
        private val factory: IntFunction<T>, private val length: ToIntFunction<T>,
        private val defaultSize: Int, private val onDiscard: Consumer<T>?, private val maxPoolSize: Int
    ) : ArrayPool<T> {
        private val pool = ConcurrentLinkedDeque<T & Any>()
        private val count = AtomicInteger()

        override fun acquire(minSize: Int): T {
            // pollFirst directly — no peek/poll TOCTOU
            val arr = pool.pollFirst()
            if (arr != null) {
                count.decrementAndGet()
                if (length.applyAsInt(arr) >= minSize) return arr
                // too small — drop it, allocate fresh
            }
            return factory.apply(max(defaultSize, minSize))
        }

        override fun release(arr: T) {
            if (arr == null) return
            val len = length.applyAsInt(arr)
            if (len < defaultSize * 0.5 || len > defaultSize * 2) {
                onDiscard?.accept(arr)
                return
            }
            var c: Int
            do {
                c = count.get()
                if (c >= maxPoolSize) {
                    onDiscard?.accept(arr)
                    return
                }
            } while (!count.compareAndSet(c, c + 1))
            pool.offerFirst(arr)
        }
    }

    enum class Strategy {
        THREAD_LOCAL,
        SHARED
    }

    class Builder<T>(
        private val factory: IntFunction<T>,
        private val length: ToIntFunction<T>
    ) {
        private var defaultSize: Int = 0
        private var onDiscard: Consumer<T>? = null
        private var maxSize: Int = -1
        private var strategy: Strategy = Strategy.THREAD_LOCAL

        fun defaultSize(defaultSize: Int) = apply { this.defaultSize = defaultSize }
        fun onDiscard(fn: Consumer<T>) = apply { this.onDiscard = fn }
        fun maxSize(maxSize: Int) = apply { this.maxSize = maxSize }
        fun strategy(strategy: Strategy) = apply { this.strategy = strategy }

        fun build(): ArrayPool<T> {
            return when (strategy) {
                Strategy.THREAD_LOCAL -> buildThreadLocal()
                Strategy.SHARED -> buildShared()
            }
        }

        private fun buildThreadLocal(): ArrayPool<T> = when {
            maxSize < 0 -> ThreadLocalPool(factory, length, defaultSize, onDiscard, Int.MAX_VALUE)
            onDiscard != null -> ThreadLocalPool(factory, length, defaultSize, onDiscard, maxSize)
            else -> ThreadLocalPool(factory, length, defaultSize, null, maxSize)
        }

        private fun buildShared(): ArrayPool<T> = when {
            maxSize < 0 -> SharedPool(factory, length, defaultSize, onDiscard, Int.MAX_VALUE)
            onDiscard != null -> SharedPool(factory, length, defaultSize, onDiscard, maxSize)
            else -> SharedPool(factory, length, defaultSize, null, maxSize)
        }
    }

    companion object {
        @JvmStatic
        fun <T> builder(factory: IntFunction<T>, length: ToIntFunction<T>): Builder<T> =
            Builder(factory, length)

        @JvmStatic
        fun <T> threadLocal(
            factory: IntFunction<T>, length: ToIntFunction<T>,
            defaultSize: Int,
            maxPerThread: Int,
            onDiscard: Consumer<T>? = null
        ): ArrayPool<T> = if (onDiscard == null) {
            Builder(factory, length)
                .defaultSize(defaultSize)
                .maxSize(maxPerThread)
                .build()
        } else {
            Builder(factory, length)
                .defaultSize(defaultSize)
                .onDiscard(onDiscard)
                .maxSize(maxPerThread)
                .build()
        }

        @JvmStatic
        fun <T> shared(
            factory: IntFunction<T>, length: ToIntFunction<T>,
            defaultSize: Int,
            maxPoolSize: Int,
            onDiscard: Consumer<T>? = null
        ): ArrayPool<T> = if (onDiscard == null) {
            Builder(factory, length)
                .strategy(Strategy.SHARED)
                .defaultSize(defaultSize)
                .maxSize(maxPoolSize)
                .build()
        } else {
            Builder(factory, length)
                .strategy(Strategy.SHARED)
                .defaultSize(defaultSize)
                .onDiscard(onDiscard)
                .maxSize(maxPoolSize)
                .build()
        }
    }
}
